ros/src/simulation/gazebo_simulator/vehicle/vehicle_gazebo_simulation_interface/src/vehicle_gazebo_input_subscriber.cpp

```diff
@@ -0,0 +1,171 @@
+#include <ros/ros.h>
+#include <geometry_msgs/Twist.h>
+#include <geometry_msgs/TwistStamped.h>
+#include <std_msgs/Float64.h>
+#include <cmath>
+#include <autoware_msgs/ControlCommandStamped.h>
+
+class VehicleGazeboInputSubscriber
+{
+  private:
+    ros::NodeHandle nh_;
+    ros::NodeHandle pnh_;
+    ros::Publisher wheel_right_rear_pub_;
+    ros::Publisher wheel_left_rear_pub_;
+    ros::Publisher steering_right_front_pub_;
+    ros::Publisher steering_left_front_pub_;
+    ros::Subscriber twiststamped_sub_;
+    ros::Subscriber twist_sub_;
+    ros::Subscriber steering_angle_sub_;
+    ros::Subscriber velocity_sub_;
+    ros::Subscriber ctrl_cmd_sub_;
+
+    void twistStampedCallback(const geometry_msgs::TwistStamped::ConstPtr &input_twist_msg);
+    void twistCallback(const geometry_msgs::Twist::ConstPtr &input_twist_msg);
+    void sterringAngleCallback(const std_msgs::Float64::ConstPtr &input_steering_angle_msg);
+    void velocityCallback(const std_msgs::Float64::ConstPtr &input_velocity_msg);
+    void controlCommandStampedCallback(const autoware_msgs::ControlCommandStamped::ConstPtr &input_msg);
+
+    double wheel_base_;
+    double wheel_tread_;
+    double wheel_radius_;
+
+  public:
+    VehicleGazeboInputSubscriber();
+    ~VehicleGazeboInputSubscriber(){};
+};
+
+VehicleGazeboInputSubscriber::VehicleGazeboInputSubscriber() : nh_(""), pnh_("~")
+{
+
+    pnh_.param("wheel_base", wheel_base_, 2.95);
+    pnh_.param("wheel_radius", wheel_radius_, 0.341);
+    pnh_.param("wheel_tread", wheel_tread_, 1.55);
+    bool twist_sub, twiststamped_sub, steering_angle_sub, velocity_sub, cmd_vel_sub, ctrl_cmd_sub;
+    pnh_.param("twist_sub", twist_sub, true);
+    pnh_.param("twiststamped_sub", twiststamped_sub, false);
+    pnh_.param("steering_angle_sub", steering_angle_sub, false);
+    pnh_.param("velocity_sub", velocity_sub, false);
+    pnh_.param("ctrl_cmd_sub", ctrl_cmd_sub, true);
+    wheel_right_rear_pub_ = nh_.advertise<std_msgs::Float64>("wheel_right_rear_velocity_controller/command", 1, true);
+    wheel_left_rear_pub_ = nh_.advertise<std_msgs::Float64>("wheel_left_rear_velocity_controller/command", 1, true);
+    steering_right_front_pub_ = nh_.advertise<std_msgs::Float64>("steering_right_front_position_controller/command", 1, true);
+    steering_left_front_pub_ = nh_.advertise<std_msgs::Float64>("steering_left_front_position_controller/command", 1, true);
+
+    if (twist_sub)
```

Comment:
These subscribers can be merged to `vehicle_cmd`, then it's better users to be able to switch to use `twist` or `ctrl` in `vehicle_cmd` by rosparam.

Reply:
38204b91ba030295cc00832809e53038c4e51091

Reply:
You want to twist command is enabled as a default, right? (`cmd_vel` is not usually used in Autoware, so no big problem. :) )